test/xpu/test_int4_linear.py

```diff
@@ -0,0 +1,205 @@
+        infeatures = weight.shape[0]
+        g_idx = torch.tensor(
+            [i // q_config["group_size"] for i in range(infeatures)],
+            dtype=torch.int32,
+        )
+        scale_zeros = gptq_zeros * gptq_scales
+        weight = gptq_scales[g_idx.long()] * weight - scale_zeros[g_idx.long()]
+        return weight
+
+    @staticmethod
+    def rand_int4(size, dtype=torch.int32, device="xpu"):
+        rand = torch.randint(-128, 128, [size // 2], device=device).to(torch.int8)
+        return rand.view(dtype=dtype)
+
+    @parametrize("per_channel", [False], lambda k: "per_channel" * k)
+    @parametrize("dtype", [torch.float16])
+    @parametrize("m,n,k", [(8, 4096, 4096), (1, 4096, 11008), (32, 4096, 4096)])
+    def test_gemm_int4(self, m, n, k, per_channel, dtype):
+        input = torch.rand([m, k], device="xpu", dtype=dtype)
+        input_torch = input.cpu()
+        weight = self.rand_int4(k * n, torch.int32, "xpu").reshape(k // 8, n)
+
+        group_size = min(128, k)
+        if per_channel:
+            group_size = k
+        group_num = int(k / group_size)
+
+        scales = torch.rand([group_num, n], device="xpu", dtype=dtype)
+        zero_points = self.rand_int4(group_num * n, torch.int32, "xpu").reshape(
+            group_num, n // 8
+        )
+
+        weight_fp = self.dequantize(
+            weight, scales, zero_points, group_size).cpu()
+        # check gemm
+        zero_points = torch.Tensor([8]).to(torch.int8).to("xpu")
+        weight_ba = weight.transpose(0, 1).contiguous()
+
+        out_onednn =torch._weight_int4pack_mm_with_scales_and_zeros(
+            input, weight_ba, scales, zero_points, group_size
+        )
+        out_torch = torch.matmul(input_torch, weight_fp)
+        self.assertEqual(
+            out_onednn.cpu().float(),
+            out_torch.float(),
+            atol=checking_atol,
+            rtol=checking_rtol,
+        )
+        # check gemm + residual
+        res0 = torch.rand([m, n], device="xpu", dtype=dtype)
+        out_onednn_res = torch._weight_int4pack_mm_with_scales_and_zeros(
+            input, weight_ba, scales, zero_points, group_size, res0)
+        out_torch_res = out_torch + res0.cpu().float()
+        self.assertEqual(
+            out_onednn_res.cpu().float(),
+            out_torch_res.float(),
+            atol=checking_atol,
+            rtol=checking_rtol,
+        )
+
+        # check gemm + bias
+        bias = torch.rand([1, n], device="xpu", dtype=dtype)
+        out_onednn_bias = torch._weight_int4pack_mm_with_scales_and_zeros(
+            input, weight_ba, bias, scales, zero_points, group_size)
+        out_torch_bias = out_torch + bias.cpu().float()
+        self.assertEqual(
+            out_onednn_bias.cpu().float(),
+            out_torch_bias.float(),
+            atol=checking_atol,
+            rtol=checking_rtol,
+        )
+
+        # check gemm + bias + gelu
+        out_onednn_gelu = torch._weight_int4pack_mm_with_scales_and_zeros(
```

Comment:
where was the signature with "tanh" defined?
does pytorch has a packed int4 gemm with post op?

Reply:
delete this UT